File: basalt/org/fenfire/view/MindMapView2D.java
/*
MindMapView2D.java
 *    
 *    Copyright (c) 2004, Matti J. Katila and Asko Soukka
 *    This file is part of Fenfire.
 *    
 *    Fenfire is free software; you can redistribute it and/or modify it under
 *    the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *    
 *    Fenfire is distributed in the hope that it will be useful, but WITHOUT
 *    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 *    Public License for more details.
 *    
 *    You should have received a copy of the GNU General
 *    Public License along with Fenfire; if not, write to the Free
 *    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *    MA  02111-1307  USA
 *    
 */
/*
 * Written by Matti J. Katila and Asko Soukka
 */

package org.fenfire.view;
import org.fenfire.view.management.*;
import org.fenfire.swamp.*;
import org.fenfire.vocab.*;
import org.fenfire.fenmm.*;
import org.fenfire.fenmm.impl.*;
import org.nongnu.libvob.*;
import org.nongnu.libvob.mouse.*;
import org.nongnu.libvob.physics.*;
import org.nongnu.libvob.physics.impl.*;

import java.util.*;

/** View representing of mind map.
 */
public class MindMapView2D extends View2D {
    private static void p(String s) { System.out.println("MMView2D:: "+s); }

    
    //ConstGraph graph = null;
    public void setGraph(ConstGraph g) {
	net = new RingMindNetImpl(g);
	geometry.construct(net);
        ((ParticleSpaceImpl)particleSpace).init();
	
    }

    public void flush() {
	actions.clear();
	particleSpace = new ParticleSpaceImpl();
	simulator = new ParticleSimulator(particleSpace);
	geometry = new GeometryImpl(simulator,
				    particleSpace,
				    new GeometryPlugin());
	geometry.construct(net);
    }

    static final Object LOCK = new Object();


    RingMindNet net;
    ParticleSpace particleSpace;
    ParticleSimulator simulator;
    Geometry geometry;
    Thread background;
    Zoomable zoo;
    WindowAnimation anim;
    final FServer f;
    public MindMapView2D(FServer f, Zoomable z, final WindowAnimation win) {
	this.f = f;
	this.zoo = z;
	this.anim = win;

	particleSpace = new ParticleSpaceImpl();
	simulator = new ParticleSimulator(particleSpace);

	geometry = new GeometryImpl(simulator,
				    particleSpace,
				    new GeometryPlugin());
	

	background = new Thread() {
		public void run() {
		    while (true) {
		    try {
			synchronized(LOCK) {
			    simulator.simulate(0.1f);
			    geometry.setCSparams(lastScene, -1);
			}
			win.rerender();
			sleep(120);
		    } catch (Exception e) { e.printStackTrace(); }
		    } 
		}
	    };
	background.start();
    }


/*
    def __init__(self, fen, context):
        self.fen = fen
        self.context = context

        self.multiplexer = self.context.multiplexer

        self.depth = 7
        self.filletWidth = 20
        self.filletLength = 150
        self.initRotation = 0
        
        self.geometry = MMGeometry(self.fen.graph, self.filletLength, self.filletWidth,
                                   self.initRotation, self.depth)

        # fillet set up
        self.angle = 1
        self.thick = 1
        self.drawEdge = 0
        self.drawInside = 1
        self.drawInsideColor = java.awt.Color.blue
        self.depthColor = 1
        self.lines = 0
        self.ellipses = 1
        self.stretched = 1
        self.curvature = 0
        self.sectors = 1
        self.fillets = 1
        self.dice = 10
        self.fillet3d = 1
        self.blend3d = 0
        self.linewidth = 1
        self.perspective = 0
        self.texture = 0
        self.dicelen = 100
        self.tblsize = 20
        self.mode = 0

        self.oldCenter = None

        self.t = None
        self.particles = None
        self.simul = None
*/

    int matchingParent = -1;

    public void render(VobScene vs, Object node,
		       int matchingParent, int box2screen, 
		       int box2plane) {

        this.matchingParent = matchingParent;

	int paper2box = vs.invertCS(box2plane, "minMap_INv");
	int paper2screen = vs.concatCS(matchingParent, "mindMap_CONCAT",
				       paper2box);

        //# XXX
        //self.zoomPanCS = paper2screen
	p("render!");
	synchronized(LOCK) {
	    lastScene = vs;
	    ((ParticleSpaceImpl)particleSpace).init();
	    net.constructNet(node);
	    geometry.construct(net);
        
	    geometry.generateCS(vs, paper2screen);

	    drawMindMap(vs);
	}
    }

    static class L implements MouseClickListener, Action.RequestHandler {
	VobScene lastScene;
	Object o;
	int cs;
	L(VobScene last, int c, Object o) {
	    lastScene = last;
	    this.o = o;
	    cs = c;
	}
	public void clicked(int x, int y) {
	    p("node!: "+lastScene.getKeyAt(cs, x,y, null));
	}
	public Object handleRequest(Object req) {
	    p("req: "+req);
	    if (req.equals("link destination")) {
		p("catch please!!!");
		return o;
	    }
	    return null;
	}
    }


    Map actions = new HashMap();
    VobScene lastScene = null;
    private void drawMindMap(VobScene vs) {
        Map links = new HashMap();


	Iterator i = net.iterateAllNodes();
	while (i.hasNext()) {
            Object node = i.next();
            Place pl = geometry.getPlace(node);
            if (pl == null)
		throw new Error(" raise 'argh! no place?!'");
            List c = new ArrayList();
	    c.add(""+pl.cs);

	    Iterator it = net.iterator(node);
            while (it.hasNext()) {
                Object n = it.next();
                Place pl2 = geometry.getPlace(n);
                if (pl2 == null)
		    throw new Error("raise 'argh! again no place?!'");
                c.add(""+pl2.cs);
	    }
	    /*    
            if dbg: p('Fillet coordinates:', c)
            def pc(conns, cs):
                if self.context.main.getPlane() == node:
                    vs.put(vob.putil.misc.getDListNocoords("Color 1 0 0"))
                vs.put(conns, cs+c)

            //# draw fillets
	    //vob.fillet.light3d.drawFillets(self, vs, pc)
	    */

            //# draw text etc..
            vs.put(new org.nongnu.libvob.vobs.RectBgVob(java.awt.Color.white), pl.cs);
            int cs = vs.coords.translate(pl.cs, 0,0, -100);
            vs.matcher.add(matchingParent, pl.cs, node);
	    vs.activate(pl.cs); // XXX

	    if (!actions.containsKey(node)) {
		L link = new L(lastScene, matchingParent, node);
		Action a = new Action(link);
		a.setListener(1,0,"click", link);
		addActions(a, node, zoo, (ParticleSpaceImpl)particleSpace, 
			   geometry.getPlace(node).particle, anim);

		Linkables.addActions(a, f, node, STRUCTLINK.linkedTo, null);
		actions.put(node, a);
	    }
	    vs.actions.put(pl.cs, (Action)actions.get(node));
	    p("put actions: "+vs+", cs: "+pl.cs);

	    int depth = geometry.getPlace(node).depth;



            it = net.iterator(node);
            while (it.hasNext()) {
                Object n = it.next();
                Place pl2 = geometry.getPlace(n);
                vs.put(new org.nongnu.libvob.vobs.SimpleConnection(
                       .5f,.5f,.5f,.5f), pl.cs, pl2.cs);

            }



	    /*
            if (depth < this.depth):
                self.putNodeContent(vs, node, cs)
            else:
                p = self.multiplexer.f(self.fen.graph, node) 
                cs = vs.orthoBoxCS(cs, node, 0, 0, 0, 0, 0, 0, 0)
                p.place(vs, cs)
	    */
	}
    }

/*
    pulic def putNodeContent(self, vs, node, into):
        # scaling after depth
        depth = self.geometry.getPlace(node).depth
        scale = ff.fenmm.MMGeometry.getScale(depth)
        textScale = ff.fenmm.MMGeometry.getTextScale(depth)

        p = self.multiplexer.f(self.fen.graph, node) 

        x = (self.filletWidth*scale)/2. - (p.getWidth()*textScale)/2.
        y = (self.filletWidth*scale)/2. - (p.getHeight()*textScale)/2.

        cs = vs.orthoBoxCS(into, node,0, x,y, textScale, textScale, p.getWidth(), p.getHeight())
        #p.place(vs, cs)

        if self.net.getLeafs().contains(node):
            
            vs.put(org.nongnu.libvob.vobs.RectBgVob(java.awt.Color.magenta), cs)
        p.place(vs, cs)


*/

    public Object getSelectedObject(Object plane, float x, float y, float w, float h) {
	return null;
    }



    public void chgFast(VobScene vs, 
			Object plane,
			int matchingParent,
			int box2screen, int box2plane) { 
        geometry.setCSparams(vs, 1);
    }







    static public void addActions(Action a, 
				  final Object ob, 
				  final Zoomable z,
				  final ParticleSpaceImpl s, 
				  final Particle p, 
				  final WindowAnimation anim) {
	a.setListener(1, 0, "move mind objects", move(ob, z, s, p, anim));
    }


    static MousePressListener move(final Object o, 
				   final Zoomable z,
				   final ParticleSpaceImpl s,
				   final Particle p, 
				   final WindowAnimation anim) {
	return new MousePressListener() {
		public MouseDragListener pressed(int x, int y) {
		    return new RelativeAdapter(){
			    boolean stagnant = false;
			    public void startDrag(int x, int y) {
				super.startDrag(x,y);
				stagnant = s.isStagnant(p);
				synchronized(LOCK) {
				s.createStagnantParticle(o,p);
				}
				p("start "+o);
			    }
			    public void endDrag(int x, int y) {
				super.endDrag(x,y);
				if (!stagnant)
				synchronized(LOCK) {
				    s.setLive(p);
				}
				p("stop");
			    }
			    public void changedRelative(float x, float y) {
				synchronized(LOCK) {
				p.x(p.x() + x/z.getZoom());
				p.y(p.y() + y/z.getZoom());
				}
				// no need..  thread updates.. 
				// anim.rerender();
			    }
			};
		}
	    };
    }    





}

File: org/fenfire/view/lava/mindMapView2D.py
# 
# Copyright (c) 2003, Matti J. Katila, Asko Soukka
# This file is part of Fenfire.
# 
# Fenfire is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Fenfire is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
# 
# You should have received a copy of the GNU General
# Public License along with Fenfire; if not, write to the Free
# Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA
# 

from __future__ import nested_scopes
import java, jarray
from java.lang import Math

import vob
import org.fenfire as ff
from org.fenfire.vocab import STRUCTLINK, RDF
from org.fenfire.vocab.lava import MINDSTRUCT
from org.fenfire.util.lava import Traversals
from org.fenfire.fenmm import MindNet, MMGeometry

def p(*s):
    print 'mindMapView2D', s

dbg = 0

class MindMapMainNode2D(ff.view.buoy.MainNode2D):
    def __init__(self, fen, centered, context, view2d, controller):
	ff.view.buoy.MainNode2D.__init__(self, centered, view2d, controller);
        self.fen = fen
        self.context = context
	self.structLink = ff.structure.StructLink.create(self.fen.graph)

        style = self.context.style

        nodeview = self.context.functional.createFunctionInstance(
            "TextNodeView",
            ff.view.TextNodeView,
            [self.context.contentFunction, style, java.awt.Color.black, 250.])

        # nodeview for normal nodes
        normalf = self.context.functional.createFunctionInstance(
            "NormalTextNodeView",
            ff.fenmm.WhiteNodeView,
            [nodeview, java.awt.Color(.6, .6, 0)])
        
        # nodeview for accursed nodes
        accursedf = self.context.functional.createFunctionInstance(
            "AccursedTextNodeView",
            ff.fenmm.WhiteNodeView,
            [nodeview, java.awt.Color(.9, .9, .9)])
            
        self.multiplexer = ff.swamp.MultiplexerNodeFunction(
            normalf.getCallableFunction(),
            accursedf.getCallableFunction())

        self.floatingView = MindMapView2D(fen, context)
        self.floatingView.multiplexer = self.multiplexer
        self.floatingView.filletLength = 40
        self.floatingView.filletWidth = 25
        self.floatingView.initRotation = 0
        self.floatingView.depth = 1
        self.floatingView.geometry.init(40, 25, 0, 1)
        
    def renderMain(self, vs, into):
        ff.view.buoy.MainNode2D.renderMain(self, vs, into)
        self.floatingView.matchingParent = into

        accursed = self.context.c.getAccursed()
        # adding accursed to multiplexer
        set = java.util.HashSet()
        if accursed:
            set.add(accursed)
            self.multiplexer.setMultiplexerNodes(set)
        else: self.multiplexer.setMultiplexerNodes(set)

        floating = self.floatingBuffer(vs, into, self.context.components)
        i = floating.keySet().iterator()

        while i.hasNext():
            node = i.next()
            if self.structLink.isLinked(node):
                self.floatingView.net = floating.get(node)
                self.floatingView.drawMindMap(vs)
            else:
                self.place(vs, node, floating.get(node))
            
    def floatingBuffer(self, vs, into, buffer):
        map = java.util.HashMap()
        if len(buffer) == 0: return map

        d = 360.0/len(buffer)
        for i in range(len(buffer)):
            if self.structLink.isLinked(buffer[i]):
                path = java.util.ArrayList(1)
                path.add(buffer[i])
                cs = self.singleCS(vs, into, buffer[i], Math.toRadians(i*d))
                cs = vs.coords.translate(cs, 0,0, 100)
                vs.matcher.add(cs, "Floating_"+buffer[i])
                map.put(buffer[i], self.floatingView.geometry.buildMindNet(vs, cs, path))
            else:
                map.put(buffer[i], self.singleCS(vs, into, buffer[i], Math.toRadians(i*d)))
                
        return map
            
    def singleCS(self, vs, into, node, angle):
        angle += 0.5 * Math.PI
        angle %= 2.0 * Math.PI
        wi, h = vs.size.width, vs.size.height
        a, b = wi/2.3, h/2.3
        x0,y0 = wi/2.0, h/2.0
        bx, by = 1, 1
        
        e = Math.sqrt(1 - b**2/a**2)

        r = a * Math.sqrt( (1 - e**2) / ( 1 - e**2 * Math.cos(angle)**2)) 
        
        x = r * Math.cos(angle) + x0
        y = -r * Math.sin(angle) + y0
         
        cs = vs.orthoBoxCS(into, node+"_FILLET", -100, x-bx/2,y-by/2, 1,1, bx,by)
        return cs

    def place(self, vs, node, into):
        p = self.multiplexer.f(self.fen.graph, node)
        x = - p.getWidth()/2.
        y = - p.getHeight()/2.
        vs.matcher.add(self.floatingView.matchingParent, into, node)
        cs = vs.orthoBoxCS(into,node,0, x, y, 1,1, p.getWidth(), p.getHeight())
        p.place(vs, cs)

#        csEnlarged = vs.coords.orthoBox(into,0, x*2, y*2, 1,1, p.getWidth()*2, p.getHeight()*2)
#        vs.matcher.add(cs, csEnlarged, "Enlarged_"+node)

#        vs.coords.activate(csEnlarged)
        vs.coords.activate(cs)

class MindMapView2D(ff.view.View2D):
    def __init__(self, fen, context):
        self.fen = fen
        self.context = context

        self.multiplexer = self.context.multiplexer

        self.depth = 5
        self.filletWidth = 20
        self.filletLength = 150
        self.initRotation = 0
        
        self.geometry = MMGeometry(self.fen.graph, self.filletLength, self.filletWidth,
                                   self.initRotation, self.depth)

        # fillet set up
        self.angle = 1
        self.thick = 1
        self.drawEdge = 0
        self.drawInside = 1
        self.drawInsideColor = java.awt.Color.blue
        self.depthColor = 1
        self.lines = 0
        self.ellipses = 1
        self.stretched = 1
        self.curvature = 0
        self.sectors = 1
        self.fillets = 1
        self.dice = 10
        self.fillet3d = 1
        self.blend3d = 0
        self.linewidth = 1
        self.perspective = 0
        self.texture = 0
        self.dicelen = 100
        self.tblsize = 20
        self.mode = 0

        self.oldCenter = None

    def render(self, vs, node,
               matchingParent, box2screen, box2plane):
        self.matchingParent = matchingParent

	paper2box = vs.invertCS(box2plane, "minMap_INv")
	paper2screen = vs.concatCS(box2screen, 'mindMap_CONCAT',
                                   paper2box)
        # XXX
        self.zoomPanCS = paper2screen
        
        path = Traversals.findShortestPath((self.oldCenter or node),
                                          STRUCTLINK.linkedTo,
                                          node, self.fen.graph)
        if not path or path.size() == 0:
            path = java.util.ArrayList(1)
            path.add(node)
         
        self.geometry.init(self.filletLength, self.filletWidth,
                           self.initRotation, self.depth)
        self.net = self.geometry.buildMindNet(vs, paper2screen, path)
        self.drawMindMap(vs)
        self.oldCenter = node

    def drawMindMap(self, vs):
        i = self.net.iterator()
        while i.hasNext():
            node = i.next()
            pl = self.net.getPlace(node)
            if pl == None: continue
            c = [ pl.cs ]

            it = self.net.iterator(node)
            while it.hasNext():
                n = it.next()
                pl2 = self.net.getPlace(n)
                if pl2 == None: continue
                c.append(pl2.cs)
                if dbg: p('info:', pl2.cs, pl2.x, pl2.y)
                
            if dbg: p('Fillet coordinates:', c)
            def pc(conns, cs):
                if self.context.main.getPlane() == node:
                    vs.put(vob.putil.misc.getDListNocoords("Color 1 0 0"))
                vs.put(conns, cs+c)

            # draw fillets
            vob.fillet.light3d.drawFillets(self, vs, pc)

            # draw text etc..
            cs = vs.coords.translate(pl.cs, 0,0, -100)
            vs.matcher.add(self.matchingParent, cs, node)

            depth = self.net.getDepth(node)
            if depth < self.depth:
                self.putNodeContent(vs, node, cs)
            else:
                p = self.multiplexer.f(self.fen.graph, node) 
                cs = vs.orthoBoxCS(cs, node, 0, 0, 0, 0, 0, 0, 0)
                p.place(vs, cs)

    def putNodeContent(self, vs, node, into):
        # scaling after depth
        depth = self.net.getDepth(node)
        scale = ff.fenmm.MMGeometry.getScale(depth)
        textScale = ff.fenmm.MMGeometry.getTextScale(depth)

        p = self.multiplexer.f(self.fen.graph, node) 

        x = (self.filletWidth*scale)/2. - (p.getWidth()*textScale)/2.
        y = (self.filletWidth*scale)/2. - (p.getHeight()*textScale)/2.

        cs = vs.orthoBoxCS(into, node,0, x,y, textScale, textScale, p.getWidth(), p.getHeight())
        p.place(vs, cs)


        ## XXX This could work, but setting TextCursor after selecting, would be difficult  
        # set the active coordinate system coverage  
#        if x > -self.filletWidth/2:  
#            x = -self.filletWidth/2
#            w = self.filletWidth*scale*2
#        else: w = p.getWidth()*textScale  
#        if y > -self.filletWidth/2:  
#            y = -self.filletWidth/2
#            h = self.filletWidth*scale*2
#        else: h = p.getHeight()*textScale  
          
#        csEnlarged = vs.coords.orthoBox(into,0, x,y, textScale, textScale, w, h)
#        vs.matcher.add(cs, csEnlarged, "Enlarged_"+node)
        
#        vs.coords.activate(csEnlarged)
        vs.coords.activate(cs)
